openmdao/core/system.py

```diff
@@ -4153,62 +4158,41 @@ def list_vars(self,
         to_remove = ['discrete']
-        if tags:
+        if not print_tags:
             to_remove.append('tags')
         if not prom_name:
             to_remove.append('prom_name')
-        for _, meta in outputs.items():
+        for _, meta in chain(outputs.items(), inputs.items()):
             for key in to_remove:
-                del meta[key]
-        for _, meta in inputs.items():
-            for key in to_remove:
-                del meta[key]
+                try:
+                    del meta[key]
+                except KeyError:
+                    pass
 
         variables = set(outputs.keys()).union(set(inputs.keys()))
         var_list = []
         var_dict = {}
 
-        real_vars = self._var_allprocs_abs2meta
-        disc_vars = self._var_allprocs_discrete
-
-        if self._subsystems_allprocs:
-            from openmdao.core.component import Component
-            for subsys in self.system_iter(recurse=True, typ=Component):
-                prefix = subsys.pathname + '.'
-                for var_name in chain(real_vars['input'], real_vars['output'],
-                                      disc_vars['input'], disc_vars['output']):
-                    if var_name in variables:
-                        if var_name.startswith(prefix):
-                            var_list.append(var_name)
-                            if var_name in outputs:
-                                var_dict[var_name] = outputs[var_name]
-                                var_dict[var_name]['io'] = 'output'
-                            else:
-                                var_dict[var_name] = inputs[var_name]
-                                var_dict[var_name]['io'] = 'input'
-        else:
-            # For components with no children, self._subsystems_allprocs is empty.
-            for var_name in chain(real_vars['input'], real_vars['output'],
-                                  disc_vars['input'], disc_vars['output']):
-                if var_name in variables:
-                    var_list.append(var_name)
-                    if var_name in outputs:
-                        var_dict[var_name] = outputs[var_name]
-                        var_dict[var_name]['io'] = 'output'
-                    else:
-                        var_dict[var_name] = inputs[var_name]
-                        var_dict[var_name]['io'] = 'input'
+        var_list = self._get_vars_exec_order(inputs=True, outputs=True,
+                                             variables=variables, local=True)
+        for var_name in var_list:
+            if var_name in outputs:
+                var_dict[var_name] = outputs[var_name]
+                var_dict[var_name]['io'] = 'output'
+            else:
+                var_dict[var_name] = inputs[var_name]
+                var_dict[var_name]['io'] = 'input'
 
         if all_procs or self.comm.rank == 0:
             write_var_table(self.pathname, var_list, 'all', var_dict,
```

Comment:
Since python dict is now ordered, we could probably get rid of the var_list arg to write_var_table and just use var_dict.

Reply:
oh good call.. forgot about that... wilco